### PR TITLE
tensorflow protobuf support

### PR DESCRIPTION
**Pull request by kahmed10**
Experimental support. Currently [inception_v3](https://github.com/tensorflow/models/tree/master/research/slim) from the tf slim models passes read, perf, and verify tests.

### CODE REVIEW

File: src/tf/tf.cpp [hunk 1]
```diff
@@ -0,0 +1,775 @@
...
+        {
+            auto new_idx = i;
+            parse_axis(new_idx);
+            new_data.at(new_idx) = prev_data.at(i);
+        }
+        prev_data = new_data;
+    }
+
+    template <class T>
+    void parse_axis(T& dim) const
```
Comment (pfultz2) [inline src/tf/tf.cpp]: Can we make this return a new dim instead of modifying it in place?

File: src/tf/tf.cpp [hunk 2]
```diff
@@ -0,0 +1,775 @@
...
+
+    node_map nodes;
+    std::vector<tensorflow::NodeDef> input_nodes;
+    std::unordered_map<std::string, instruction_ref> instructions;
+    program prog = program();
+    bool is_nhwc = true;
+
+    std::unordered_map<std::string, op_func> ops;
+
+    std::vector<size_t> parse_axes(attribute_map& attributes, const std::string& s) const
```
Comment (pfultz2) [inline src/tf/tf.cpp]: Can the `attribute_map` be constant?

File: src/tf/tf.cpp [hunk 3]
```diff
@@ -0,0 +1,775 @@
...
+            const std::string& pad_mode = attributes.at("padding").s();
+            if(pad_mode.find("SAME") != std::string::npos)
+            {
+                op.padding_mode = op::padding_mode_t::same;
+            }
+            else if(pad_mode.find("EXPLICIT") != std::string::npos)
+            {
+                std::vector<size_t> padding;
+                copy(attributes.at("explicit_paddings").list().i(), std::back_inserter(padding));
+                reorder_data(padding);
```
Comment (pfultz2) [inline src/tf/tf.cpp]: Is this necessary? Padding is not in a nchw format. Its in the format of something like `H1, H2, W1, W2`.

File: src/tf/tf.cpp [hunk 4]
```diff
@@ -0,0 +1,775 @@
...
+        if(weights->name() != "@param")
+        {
+            if(is_nhwc)
+            {
+                weights = prog.add_instruction(op::transpose{{1, 3, 0, 2}}, args[1]);
+            }
+            else
+            {
+                weights = prog.add_instruction(op::transpose{{3, 2, 0, 1}}, args[1]);
+            }
```
Comment (pfultz2) [inline src/tf/tf.cpp]: I wonder if the `reorder_data` function could be used here as well. I am not sure.

File: src/tf/tf.cpp [hunk 5]
```diff
@@ -0,0 +1,775 @@
...
+        return prog.add_instruction(op, args[0]);
+    }
+
+    instruction_ref
+    parse_reshape(const std::string&, const attribute_map&, std::vector<instruction_ref> args)
+    {
+        op::reshape op;
+        if(args.size() != 2)
+            MIGRAPHX_THROW("reshape needs 2 arguments (input, new_shape)");
+        literal s = args[1]->get_literal();
```
Comment (pfultz2) [inline src/tf/tf.cpp]: `eval` should probably be used instead of `get_literal`.

File: src/tf/tf.cpp (hunk 4 above)
Comment (kahmed10) [reply src/tf/tf.cpp]: I think this permutation is very specific to convolutions, so I will leave it as is unless we see more operators with specific permutations.

**Comment by pfultz2** [comment]
There is this error on jenkins:

```
test/tf/tf_test.cpp:87:
    FAILED: p == prog  [ @0 = @literal{1} -> int32_type, {1}, {0}
1 = @param:1 -> float_type, {4, 2, 3}, {6, 3, 1}
0 = @param:0 -> float_type, {4, 7, 3}, {21, 3, 1}
@3 = concat(0,1) -> float_type, {4, 9, 3}, {27, 3, 1}
 == @0 = @literal{1} -> int32_type, {1}, {1}
1 = @param:1 -> float_type, {4, 2, 3}, {6, 3, 1}
0 = @param:0 -> float_type, {4, 7, 3}, {21, 3, 1}
@3 = concat(0,1) -> float_type, {4, 9, 3}, {27, 3, 1}
 ]
```

Which has to do with a change to make the stride zero when making a scalar literal. The first lines differ based on the stride only:

```
@0 = @literal{1} -> int32_type, {1}, {0}
@0 = @literal{1} -> int32_type, {1}, {1}
```

**Comment by kahmed10** [comment]
This issue is because in tf.cpp:parse_tensor(), we use a different constructor for the literal, and the size of the data container happens to be 1. One way to fix this is to adjust the constructor in my tests accordingly.

**Pull request merged**